File: app/src/main/java/com/orgzly/android/repos/DropboxRepo.java
package com.orgzly.android.repos;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;

import com.orgzly.android.BookName;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;

public class DropboxRepo implements SyncRepo {
    public static final String SCHEME = "dropbox";

    private final Uri repoUri;
    private final DropboxClient client;

    public DropboxRepo(RepoWithProps repoWithProps, Context context) {
        this.repoUri = Uri.parse(repoWithProps.getRepo().getUrl());
        this.client = new DropboxClient(context, repoWithProps.getRepo().getId());
    }

    @Override
    public boolean isConnectionRequired() {
        return true;
    }

    @Override
    public boolean isAutoSyncSupported() {
        return false;
    }

    @Override
    public Uri getUri() {
        return repoUri;
    }

    @Override
    public List<VersionedRook> getBooks() throws IOException {
        RepoIgnoreNode ignores = new RepoIgnoreNode(this);
        return client.getBooks(repoUri, ignores);
    }

    @Override
    public VersionedRook retrieveBook(String repoRelativePath, File file) throws IOException {
        return client.download(repoUri, repoRelativePath, file);
    }

    @Override
    public InputStream openRepoFileInputStream(String repoRelativePath) throws IOException {
        return client.streamFile(repoUri, repoRelativePath);
    }

    @Override
    public VersionedRook storeBook(File file, String repoRelativePath) throws IOException {
        return client.upload(file, repoUri, repoRelativePath);
    }

    @Override
    public VersionedRook renameBook(Uri oldFullUri, String newName) throws IOException {
        BookName oldBookName = BookName.fromRepoRelativePath(BookName.getRepoRelativePath(repoUri, oldFullUri));
        String newRelativePath = BookName.repoRelativePath(newName, oldBookName.getFormat());
        String newEncodedRelativePath = Uri.encode(newRelativePath, "/");
        Uri newFullUri = repoUri.buildUpon().appendEncodedPath(newEncodedRelativePath).build();
        return client.move(repoUri, oldFullUri, newFullUri);
    }

    @Override
    public void delete(Uri uri) throws IOException {
        client.delete(uri.getPath());
    }

    /**
     * Only used by tests. The delete() method does not allow deleting directories.
     */
    public void deleteDirectory(Uri uri) throws IOException {
        client.deleteFolder(uri.getPath());
    }

    @NonNull
    @Override
    public String toString() {
        return repoUri.toString();
    }
}

File: app/src/main/java/com/orgzly/android/repos/DropboxRepo.kt
package com.orgzly.android.repos

import android.content.Context
import android.net.Uri
import com.orgzly.R
import com.orgzly.android.App
import com.orgzly.android.BookName
import com.orgzly.android.prefs.AppPreferences
import com.orgzly.android.ui.note.NoteAttachmentData
import com.orgzly.android.util.UriUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import kotlin.Throws

class DropboxRepo(repoWithProps: RepoWithProps, context: Context?) : SyncRepo {
    private val repoUri: Uri
    private val client: DropboxClient

    override fun isConnectionRequired(): Boolean {
        return true
    }

    override fun isAutoSyncSupported(): Boolean {
        return false
    }

    override fun getUri(): Uri {
        return repoUri
    }

    @Throws(IOException::class)
    override fun getBooks(): List<VersionedRook> {
        val ignores = RepoIgnoreNode(this)
        return client.getBooks(repoUri, ignores)
    }

    @Throws(IOException::class)
    override fun retrieveBook(repoRelativePath: String, file: File): VersionedRook {
        return client.download(repoUri, repoRelativePath, file)
    }

    @Throws(IOException::class)
    override fun openRepoFileInputStream(repoRelativePath: String): InputStream {
        return client.streamFile(repoUri, repoRelativePath)
    }

    @Throws(IOException::class)
    override fun storeBook(file: File, repoRelativePath: String): VersionedRook {
        val context = App.getAppContext()
        if (repoRelativePath.contains("/") && !AppPreferences.subfolderSupport(context)) {
            throw IOException(context.getString(R.string.subfolder_support_disabled))
        }
        return client.upload(file, repoUri, repoRelativePath)
    }

    @Throws(IOException::class)
    override fun storeFile(file: File, pathInRepo: String, fileName: String): VersionedRook {
        if (!file.exists()) {
            throw FileNotFoundException("File $file does not exist")
        }

        val relativePath = if (pathInRepo.isNotEmpty()) "$pathInRepo/$fileName" else fileName
        return client.upload(file, repoUri, relativePath)
    }

    override fun listFilesInPath(pathInRepo: String?): MutableList<NoteAttachmentData> {
        // TODO: Implement listFilesInPath.
        return mutableListOf()
    }

    @Throws(IOException::class)
    override fun renameBook(oldFullUri: Uri, newName: String): VersionedRook {
        val context = App.getAppContext()
        if (newName.contains("/") && !AppPreferences.subfolderSupport(context)) {
            throw IOException(context.getString(R.string.subfolder_support_disabled))
        }
        val oldBookName = BookName.fromRepoRelativePath(BookName.getRepoRelativePath(repoUri, oldFullUri))
        val newRelativePath = BookName.repoRelativePath(newName, oldBookName.format)
        val newEncodedRelativePath = Uri.encode(newRelativePath, "/")
        val newFullUri = repoUri.buildUpon().appendEncodedPath(newEncodedRelativePath).build()
        return client.move(repoUri, oldFullUri, newFullUri)
    }

    @Throws(IOException::class)
    override fun delete(uri: Uri) {
        client.delete(uri.path)
    }

    override fun toString(): String {
        return repoUri.toString()
    }

    companion object {
        const val SCHEME = "dropbox"
    }

    override fun getUriForPath(path: String): Uri? {
        // TODO: Implement getUriForPath.
        return null
    }

    init {
        repoUri = Uri.parse(repoWithProps.repo.url)
        client = DropboxClient(context, repoWithProps.repo.id)
    }
}
